doc/lenses.rst

```diff
@@ -3,16 +3,496 @@
 Lenses
 ======
 
-.. admonition:: Work in progress
-   :class: danger
+A tool borrowed from functional programming. While lenses are used
+extensively in Lager for zooming on cursors, they are also useful on
+their own for manipulating immutable data.
 
-   This section of the documentation has not been written yet.
+.. _making-a-lens:
 
-   Meanwhile, you can watch the **C++ Russia 2019 Talk**: `Squaring
-   the circle: value oriented design in an object oriented system
-   <https://www.youtube.com/watch?v=e2-FRFEx8CA>`_ (`slides`_).
+Making a lens
+-------------
 
-.. _slides: https://sinusoid.es/talks/cpprussia19-piter
+Lenses are, conceptually, a pair of functions for focusing on a part
+of a whole. You use a lens with the following interface:
 
-.. image:: http://orvex.org.p11.hostingprod.com/images/under_construction.gif
+.. code-block:: c++
+
+   // pseudocode for the lens interface:
+   part view(lens<whole, part>, whole); // get part of a whole
+   whole set(lens<whole, part>, whole, part); // set the part of a whole
+   whole over(lens<whole, part>, whole, std::function<part(part)>); // update the part of a whole
+   lens<A, C> compose(lens<A, B>, lens<B, C>); // compose two lenses into another lens
+
+Let's use this mouse as our model:
+
+.. image:: _static/mouse.svg
+   :width: 50%
    :align: center
+
+.. code-block:: c++
+
+   struct mouse {
+       mouth mouth;
+       pair<eye, eye> eyes;
+       vector<leg> legs;
+       vector<whisker> whiskers;
+       tail tail;
+   };
+
+So, how do we make a lens, say, to access the eyes of a mouse?
+
+.. code-block:: c++
+
+   auto eyes = [](auto f) {
+       return [f](mouse mouse) {
```

Comment:
Where you talking about this? I am happy with `mouse m`

Reply:
`mouse m` is fine, I'm more worried about struct members.